File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/typenodes/basic_and_wrapped/IBasicAndWrappedTypeNode.java
package org.vanautrui.languages.compiler.parsing.astnodes.typenodes.basic_and_wrapped;

import org.vanautrui.languages.compiler.parsing.IASTNode;
import org.vanautrui.languages.compiler.parsing.astnodes.typenodes.ITypeNode;

public interface IBasicAndWrappedTypeNode extends IASTNode, ITypeNode {

	//classes implementing this interface are also used to communicate type information throughout the compiler.
	//so it is not just for parsing.

	String getTypeName();
}

File: dragon-parser/src/main/compiler/parsing/astnodes/typenodes/basic_and_wrapped/IBasicAndWrappedTypeNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers

@JsonTypeInfo(
		use = JsonTypeInfo.Id.NAME,
		include = JsonTypeInfo.As.PROPERTY,
		property = "type")
@JsonSubTypes({
		@JsonSubTypes.Type(value = SimpleTypeNode.class, name = "SimpleTypeNode"),
		@JsonSubTypes.Type(value = SubroutineTypeNode.class, name = "SubroutineTypeNode")
})
public interface IBasicAndWrappedTypeNode extends IASTNode, ITypeNode {

	//classes implementing this interface are also used to communicate type information throughout the compiler.
	//so it is not just for parsing.
}
